### PR TITLE
[Model Control Plane] Add CLI

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
*I moved around model registry files to free up `model` name for Model WatchTower, so number of changed lines might look scary, but it is way less*

Supported CLI commands:
- `zenml model list` - list models with filter
- `zenml model register -n name -d description ...` - register model
- `zenml model delete <model_name> -y` - delete model
- `zenml model update <model_name> -d description ...` - update model
- `zenml model version list <model_name>` - list versions of a model
- `zenml model version update <model_name> <model_version> --stage production --force` - move model version to stage
- `zenml model version artifacts <model_name> <model_version>` - list all artifacts of a model version
- `zenml model version model_objects <model_name> <model_version>` - list all model_objects of a model version
- `zenml model version deployments <model_name> <model_version>` - list all deployments of a model version
- `zenml model version runs <model_name> <model_version>` - lst all pipeline runs of a model version

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/enums.py [hunk 1]
```diff
@@ -144,6 +144,7 @@ class CliCategories(StrEnum):
     HUB = "ZenML Hub"
     INTEGRATIONS = "Integrations"
     MANAGEMENT_TOOLS = "Management Tools"
+    MODEL_WATCHTOWER = "Model Watch Tower"
```
Comment (strickvl) [inline src/zenml/enums.py:147]:
```suggestion
    MODEL_WATCHTOWER = "Model Watchtower"
```

That said, we should probably agree / align on the canonical way for us to refer to this so it's standardised within ZenML...

File: src/zenml/models/model_models.py [hunk 2]
```diff
@@ -157,6 +157,42 @@ def pipeline_runs(self) -> Dict[str, PipelineRunResponseModel]:
...
+            The amount of deployments linked to this model version.
+        """
+        return len(self.deployment_ids)
+
+    @property
+    def pipeline_runs_count(self) -> int:
+        """Get the amount of pipeline runs linked to this model version.
+
+        Returns:
+            The amount of pipeline runs linked to this model version.
```
Comment (strickvl) [inline src/zenml/models/model_models.py:192]:
```suggestion
            The number of model objects linked to this model version.
        """
        return len(self.model_object_ids)

    @property
    def artifact_objects_count(self) -> int:
        """Get the number of artifact objects linked to this model version.

        Returns:
            The number of artifact objects linked to this model version.
        """
        return len(self.artifact_object_ids)

    @property
    def deployments_count(self) -> int:
        """Get the number of deployments linked to this model version.

        Returns:
            The number of deployments linked to this model version.
        """
        return len(self.deployment_ids)

    @property
    def pipeline_runs_count(self) -> int:
        """Get the number of pipeline runs linked to this model version.

        Returns:
            The number of pipeline runs linked to this model version.
```

File: tests/integration/functional/cli/model/conftest.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+#  Copyright (c) ZenML GmbH 2022. All Rights Reserved.
```
Comment (strickvl) [inline tests/integration/functional/cli/model/conftest.py:1]:
```suggestion
#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
```

File: tests/integration/functional/cli/model/test_model.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+def test_model_list(clean_workspace_with_models):
+    """Test that zenml model list does not fail."""
+    runner = CliRunner()
+    list_command = cli.commands["model"].commands["list"]
+    result = runner.invoke(list_command)
+    assert result.exit_code == 0
+
+
+def test_model_create_short_names(clean_workspace_with_models):
+    """Test that zenml model create does not fail."""
```
Comment (strickvl) [inline tests/integration/functional/cli/model/test_model.py:34]:
```suggestion
    """Test that zenml model create does not fail with short names."""
```

File: tests/integration/functional/cli/model/test_model.py [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+            "j",
+            "-t",
+            "k",
+        ],
+    )
+    assert result.exit_code == 0
+
+
+def test_model_create_full_names(clean_workspace_with_models):
+    """Test that zenml model create does not fail."""
```
Comment (strickvl) [inline tests/integration/functional/cli/model/test_model.py:68]:
```suggestion
    """Test that zenml model create does not fail with full names."""
```

File: tests/integration/functional/cli/test_model_registry.py
Comment (strickvl) [inline tests/integration/functional/cli/test_model_registry.py:1]: What's this file?

File: src/zenml/cli/model.py [hunk 6]
```diff
@@ -11,650 +11,504 @@
...
+    """List, create or delete models in the Model WatchTower."""
 
 
-def register_model_registry_subcommands() -> None:  # noqa: C901
-    """Registers CLI subcommands for the Model Registry."""
-    model_registry_group = cast(TagGroup, cli.commands.get("model-registry"))
-    if not model_registry_group:
+@cli_utils.list_options(ModelFilterModel)
+@model.command("list", help="List all models.")
+def list_models(**kwargs: Any) -> None:
```
Comment (strickvl) [inline src/zenml/cli/model.py:48]: really `Any`? Wouldn't it always be strings?

File: src/zenml/cli/model.py [hunk 7]
```diff
@@ -11,650 +11,504 @@
...
+    "--name",
+    "-n",
+    help="The name of the model.",
+    type=str,
+    required=True,
+)
+@click.option(
+    "--license",
+    "-l",
+    help="The license model created under.",
```
Comment (strickvl) [inline src/zenml/cli/model.py:78]:
```suggestion
    help="The license under which the model is created.",
```

File: src/zenml/cli/model.py [hunk 8]
```diff
@@ -11,650 +11,504 @@
...
+    "--description",
+    "-d",
+    help="The description of the model.",
+    type=str,
+    required=False,
+)
+@click.option(
+    "--audience",
+    "-a",
+    help="The target audience of the model.",
```
Comment (strickvl) [inline src/zenml/cli/model.py:92]:
```suggestion
    help="The target audience for the model.",
```

File: src/zenml/cli/model.py [hunk 9]
```diff
@@ -11,650 +11,504 @@
...
+@click.option(
+    "--ethical",
+    "-e",
+    help="The ethical implications of the model.",
+    type=str,
+    required=False,
+)
+@click.option(
+    "--limitations",
+    help="The know limitations of the model.",
```
Comment (strickvl) [inline src/zenml/cli/model.py:118]:
```suggestion
    help="The known limitations of the model.",
```

File: src/zenml/cli/model.py [hunk 10]
```diff
@@ -11,650 +11,504 @@
...
+    required=False,
+    multiple=True,
+)
+def create_model(
+    name: str,
+    license: Optional[str],
+    description: Optional[str],
+    audience: Optional[str],
+    use_cases: Optional[str],
+    trade_offs: Optional[str],
```
Comment (strickvl) [inline src/zenml/cli/model.py:136]:
```suggestion
    tradeoffs: Optional[str],
```

File: src/zenml/cli/model.py [hunk 11]
```diff
@@ -11,650 +11,504 @@
...
-            model_version = model_registry.get_model_version(name, version)
-        except KeyError:
-            cli_utils.error(
-                f"Model with name {name} and version {version} does not exist."
-            )
-            return
-        cli_utils.pretty_print_model_version_details(model_version)
+@model.group
+def version() -> None:
+    """List or view model versions in the Model WatchTower."""
```
Comment (strickvl) [inline src/zenml/cli/model.py:217]:
```suggestion
    """List or view model versions in the Model Watchtower."""
```

File: src/zenml/cli/model.py [hunk 12]
```diff
@@ -11,650 +11,504 @@
...
+@click.option(
+    "--only-model-objects",
+    "-m",
+    is_flag=True,
+    help="Show only model objects.",
+)
+@click.option(
+    "--only-deployments",
+    "-d",
+    is_flag=True,
```
Comment (strickvl) [inline src/zenml/cli/model.py:363]:
```suggestion
    is_flag=True,
    default=False,
    help="Show only artifact objects.",
)
@click.option(
    "--only-model-objects",
    "-m",
    is_flag=True,
    default=False,
    help="Show only model objects.",
)
@click.option(
    "--only-deployments",
    "-d",
    is_flag=True,
    default=False,
```

File: src/zenml/cli/model_registry.py [hunk 13]
```diff
@@ -0,0 +1,660 @@
...
+        type=str,
+        default=None,
+        help="Description of the model to update.",
+    )
+    @click.option(
+        "--metadata",
+        "-t",
+        type=(str, str),
+        default=None,
+        help="Metadata or Tags to add to the model. Ran be used like: -m key1 value1 -m key2 value",
```
Comment (strickvl) [inline src/zenml/cli/model_registry.py:216]:
```suggestion
        help="Metadata or Tags to add to the model. Can be used like: -m key1 value1 -m key2 value",
```

File: src/zenml/cli/model_registry.py [hunk 14]
```diff
@@ -0,0 +1,660 @@
...
+        "-m",
+        type=(str, str),
+        default=None,
+        help="Metadata or Tags to add to the model. can be used like: --m key1 value1 -m key2 value",
+        multiple=True,
+    )
+    @click.option(
+        "--stage",
+        "-s",
+        type=click.Choice(["None", "Staging", "Production", "Archived"]),
```
Comment (strickvl) [inline src/zenml/cli/model_registry.py:406]: No other possible choices people might want to have for stages custom to their particular process?

File: src/zenml/cli/model_registry.py [hunk 15]
```diff
@@ -0,0 +1,660 @@
...
+        type=str,
+        default=None,
+        help="Model URI of the model to list versions for.",
+    )
+    @click.option(
+        "--metadata",
+        "-m",
+        type=(str, str),
+        default=None,
+        help="Metadata or Tags to filter the model versions by. can be used like: -m key1 value1 -m key2 value",
```
Comment (strickvl) [inline src/zenml/cli/model_registry.py:481]:
```suggestion
        help="Metadata or Tags to filter the model versions by. Can be used like: -m key1 value1 -m key2 value",
```

File: src/zenml/cli/model_registry.py [hunk 16]
```diff
@@ -0,0 +1,660 @@
...
+        type=str,
+        default=None,
+        help="Description of the model version.",
+    )
+    @click.option(
+        "--metadata",
+        "-m",
+        type=(str, str),
+        default=None,
+        help="Metadata or Tags to add to the model version. can be used like: -m key1 value1 -m key2 value",
```
Comment (strickvl) [inline src/zenml/cli/model_registry.py:567]:
```suggestion
        help="Metadata or Tags to add to the model version. Can be used like: -m key1 value1 -m key2 value",
```

**Review by strickvl — CHANGES_REQUESTED** [review]

File: src/zenml/cli/model_registry.py (hunk 14 above)
Comment (avishniakov) [reply src/zenml/cli/model_registry.py:406]: This is not my code, just renamed the file to make more sense and free up `model.py` available 🙂

File: src/zenml/cli/utils.py [hunk 17]
```diff
@@ -317,6 +319,9 @@ def __dictify(model: T) -> Dict[str, str]:
             else:
                 items[k] = str(value)
         # prepend an active marker if a function to mark active was passed
+        if suppress_active_column:
+            return items
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/utils.py:323]: It is same, I just don't add extra ACTIVE column

File: tests/integration/functional/cli/test_model_registry.py
Comment (avishniakov) [reply tests/integration/functional/cli/test_model_registry.py:1]: Also, just a renaming to free up `test_model.py` reflecting new cli files structure

File: src/zenml/cli/model.py [hunk 18]
```diff
@@ -11,650 +11,507 @@
...
+    help="The ethical implications of the model.",
+    type=str,
+    required=False,
+)
+@click.option(
+    "--limitations",
+    help="The known limitations of the model.",
+    type=str,
+    required=False,
+)
```
Comment (fa9r) [inline src/zenml/cli/model.py:121]: Actually, aren't all of these simply tags? If so, are they handled in a special way in the UI? Or why are they separate args?

**Review by fa9r — CHANGES_REQUESTED** [review]
New CLI commands all look good 👍 

High-level I was wondering whether it makes sense to have both `zenml models ...` and `zenml model-registry models ...` now. Isn't this redundant?

Aside from that just a few minor comments, looks like it's ready to merge soon 👍

File: src/zenml/cli/model_registry.py
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/model_registry.py:1]: Don't they interact with the Model Registry component, which we do not plan to depreciate? Why this should be deleted?
To me, it looks aligned now.

File: src/zenml/cli/utils.py [hunk 19]
```diff
@@ -253,6 +253,7 @@ def print_pydantic_models(
     exclude_columns: Optional[List[str]] = None,
     active_models: Optional[List[T]] = None,
     show_active: bool = False,
+    suppress_active_column: bool = False,
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/utils.py:256]: There is an auto-added ACTICE column, but with active_models one can set there to post ticks in that column. Didn't want to break the backward compatibility of other CLI commands, so just added an extra param here.

File: src/zenml/cli/model.py (hunk 18 above)
Comment (avishniakov) [reply src/zenml/cli/model.py:121]: Yeah, they will be represented in the main tab of the model and model version separately, so they are structured tags in a sense.

File: src/zenml/cli/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/utils.py:256]: Will rework this

File: src/zenml/cli/__init__.py [hunk 20]
```diff
@@ -1557,3 +1557,4 @@ def my_pipeline(...):
 from zenml.cli.user_management import *  # noqa
 from zenml.cli.version import *  # noqa
 from zenml.cli.downgrade import *  # noqa
+from zenml.cli.model import *  # noqa
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/__init__.py:1560]: Will rework this

File: src/zenml/models/model_models.py [hunk 21]
```diff
@@ -157,6 +157,42 @@ def pipeline_runs(self) -> Dict[str, PipelineRunResponseModel]:
...
+
+    @property
+    def pipeline_runs_count(self) -> int:
+        """Get the number of pipeline runs linked to this model version.
+
+        Returns:
+            The number of pipeline runs linked to this model version.
+        """
+        return len(self.pipeline_run_ids)
+
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/models/model_models.py]: Done

File: src/zenml/cli/utils.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/utils.py]: Done

File: src/zenml/cli/__init__.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (avishniakov) [reply src/zenml/cli/__init__.py]: Done

**Review by strickvl — APPROVED** [review]

**Pull request merged**